modules/seqimporter/src/main/scala/gem/seqimporter/ConfigReader.scala

```diff
@@ -0,0 +1,279 @@
+package gem.seqimporter
+
+import edu.gemini.spModel.core._
+import gem._
+import gem.config._
+import gem.enum._
+
+import java.time.Duration
+
+import scala.reflect.runtime.universe.TypeTag
+import scalaz._
+import Scalaz._
+
+object ConfigReader {
+  type Read[A] = String => A
+
+  // Clean up classnames in toString, which tells you the type of the key, which should be
+  // helpful for debugging.
+  private def clean(s: String) =
+    s.replace("edu.gemini.spModel.core.", "")
+     .replace("java.lang.", "")
+     .replace("java.time.", "")
+     .replace("gem.", "")
+
+  object Read {
+    val int:     Read[Int     ] = _.toInt
+    val long:    Read[Long    ] = _.toLong
+    val double:  Read[Double  ] = _.toDouble
+    val boolean: Read[Boolean ] = _.toBoolean
+    val yesNo:   Read[Boolean ] = {
+      case "No"  => false
+      case "Yes" => true
+      case x     => sys.error(s"unrecognized Yes/No value: $x")
+    }
+    def arcsec:  Read[Angle   ] = double andThen Angle.fromArcsecs
+    val durSecs: Read[Duration] = double andThen { d => Duration.ofMillis((d * 1000).round) }
+  }
+
+  sealed abstract class System(val system: String) {
+
+    final class Key[A](val name: String, val read: Read[A])(implicit ev: TypeTag[A]) {
+      val path = s"$system:$name"
+
+      def legacyGet(cm: ConfigMap): Option[String] =
+        cm.lookup(path)
+
+      override def toString: String =
+        s"Key[${clean(ev.tpe.toString)}]($path)"
+    }
+
+    protected object Key {
+      def apply[A: TypeTag](name: String)(read: Read[A]): Key[A] =
+        new Key(name, read)
+
+      def enum[A: TypeTag](name: String, m: (String, A)*): Key[A] =
+        enumWithDefault[A](name, s => sys.error(s"new value of '$name' for $s not found"), m: _*)
+
+      def enumWithDefault[A: TypeTag](name: String, default: String => A, m: (String, A)*): Key[A] =
+        new Key[A](name, m.toMap.withDefault(default))
+    }
+  }
+
+  object Legacy {
+    case object Telescope extends System("telescope") {
+      val P = Key[OffsetP]("p")(Read.arcsec andThen OffsetP.apply)
+      val Q = Key[OffsetQ]("q")(Read.arcsec andThen OffsetQ.apply)
+    }
+
+    case object Observe extends System("observe") {
+      val ObserveType  = Key[String  ]("observeType" )(identity    )
+      val ExposureTime = Key[Duration]("exposureTime")(Read.durSecs)
+      val Status       = Key[Boolean ]("status"      ) {
+        case "complete" => true
+        case _          => false
```

Comment:
It's a bit lazy to use a `Boolean` for this.  I'll either replace it with an actual type or else, if we write a more complete ODB importer as I propose, do away with this in favor of writing out dataset information directly from the obslog.